File: Binary Search/240.search_a_2D_matrix_II.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
// Binary Search解法
var searchMatrix = function(matrix, target) {
    if (matrix === null || matrix.length === 0 || matrix[0].length === 0) return false;

    for (let i = 0; i < matrix.length; i++) {
        let diagonal = matrix[i][i];

        if (target > diagonal) {
            if (searchRow(i, target, matrix, i) || searchCol(i, target, matrix, i)) return true;
        }
    }

    return false;
};

let searchRow = function(start, target, matrix, row) {
    let left = start;
    let right = matrix[0].length;
    let arr = matrix[row];

    while (left + 1 < right) {
        let mid = Math.floor((left + right) / 2);

        if (arr[mid] === target) return true;

        else if (arr[mid] > target) {
            right = mid;
        } else {
            left = mid;
        }
    }

    if (arr[left] === target || arr[right] === target) return true;
    return false;
}


let searchRow = function(start, target, matrix, col) {
    let left = start;
    let right = matrix.length;

    let arr = [];
    for (let i = 0; i < matrix.length; i++) {
        arr.push(matrix[i][col]);
    }

    while (left + 1 < right) {
        let mid = Math.floor((left + right) / 2);

        if (arr[mid] === target) return true;

        else if (arr[mid] > target) {
            right = mid;
        } else {
            left = mid;
        }
    }

    if (arr[left] === target || arr[right] === target) return true;
    return false;
}

// Greedy Search
var searchMatrix = function(matrix, target) {
    let col = 0;
    let row = matrix.length - 1;

    while (row >= 0 && col < matrix[0].length) {

        if (matrix[row][col] > target) {
            row--;
        } else if (matrix[row][col] < target) {
            col++;
        } else {
            return true;
        }
    }

    return false;

}

File: Java/240.search_a_2D_matrix_II.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) return false;

        int shortSide = Math.min(matrix.length, matrix[0].length);

        for (int i = 0; i < shortSide; i++) {
            if (matrix[i][i] == target) {
                return true;
            } else if (matrix[i][i] < target) {
                // 这里在边界条件处理的不好，一是一开始是用的i + 1, i- 1，有可能出现OUT OF BOUND的问题
                // 二是把search ROW 和 COL的最大值边界搞反了，INTERVIEW的时候一定会出问题
                boolean res = searchInCol(matrix, i, matrix.length - 1, i, target)
                        || searchInRow(matrix, i, matrix[0].length - 1, i, target);
                if (res) return true;
            } else {
                boolean res = searchInCol(matrix, 0, i, i, target)
                        || searchInRow(matrix, 0, i, i, target);
                if (res) return true;
            }
        }
        return false;
    }

    public boolean searchInRow(int[][] matrix, int start, int end, int row, int target) {
        while (start + 1 < end) {
            int middle = (start + end) / 2;
            if (matrix[row][middle] == target) return true;
            else if (matrix[row][middle] < target) {
                start = middle + 1;
            } else {
                end = middle;
            }
        }

        if (matrix[row][start] == target || matrix[row][end] == target) return true;
        return false;
    }

    public boolean searchInCol(int[][] matrix, int start, int end, int col, int target) {
        while (start + 1 < end) {
            int middle = (start + end) / 2;
            if (matrix[middle][col] == target) return true;
            else if (matrix[middle][col] < target) {
                start = middle + 1;
            } else {
                end = middle;
            }
        }
        if (matrix[start][col] == target || matrix[end][col] == target) return true;
        return false;
    }
}